README.md
# MATLAB Integration with BeamNG.tech

<a name="About"></a>

## About

We are excited to announce that the highly requested feature of bridging BeamNG.tech and [MATLAB](https://www.mathworks.com/products/matlab.html) is here. MATLAB, with its long history as an academic engineering and mathematical tool, is a programming and numeric computing platform used to analyse data, develop algorithms, and create models. The newly created bridge with MATLAB will enable you to run, control, and interact with the BeamNG.tech simulation. We have integrated five main scripts for your convenience in making use of annotations, bounding boxes, multi-shot camera, object placement, vehicle state plotting and creation of simple scenarios on our East Coast USA map.

## Table of Contents

 - [Documentation](#docs)
 - [Features](#features) 
 - [Prerequirements](#prereqs)
 - [Setup](#setup)
 - [Getting Started](#getstart)
 - [Compatibility](#compatibility)
 - [License](#license)
 
<a name="docs"></a>

## Documentation
[![](https://raw.githubusercontent.com/BeamNG/BeamNG-MATLAB-integration/main/media/readthedocs.png)](https://beamngpy.readthedocs.io/en/latest/MATLAB.html)

<a name="features"></a>

## Features


As of now the BeamNG ROS integration supports one package for the **remote** control of the simulation platform and one package for the control of a driving agent. A third package manages custom messages.

<a name="prereqs"></a>
## Prerequirements
The [BeamNG-MATLAB bridge](https://github.com/BeamNG/BeamNG-MATLAB-integration) is depending on [BeamNG.Tech](https://documentation.beamng.com/beamng_tech/) and [BeamNGpy](https://documentation.beamng.com/beamng_tech/beamngpy/). Make sure that you have the license for BeamNG.Tech. The Github repository of the BeamNG-MATLAB has some basic examples of scripts that run a vehicle with some sensors ex. Lidar, Camera, and state sensor. 

you must have the following softwares/packages installed 
* [Compatible Python](https://www.mathworks.com/support/requirements/python-compatibility.html) 
* [BeamNGpy](https://pypi.org/project/beamngpy/)
* [BeamNG.Tech](https://documentation.beamng.com/beamng_tech/)



<a name="setup"></a>

## Setup

### 1. Setup a compatible python version   
After installing the [compatible python version](https://www.mathworks.com/support/requirements/python-compatibility.html) with MATLAB, make sure to include the path of excutable python file (exe) in your in "path" variable of "environment variables" as explained [here](https://docs.oracle.com/en/database/oracle/machine-learning/oml4r/1.5.1/oread/creating-and-modifying-environment-variables-on-windows.html#GUID-DD6F9982-60D5-48F6-8270-A27EC53807D0).  

### 2. Run pyhton engine in MATLAB 
Run the [test_python.m](https://github.com/BeamNG/BeamNG-MATLAB-integration/blob/main/test_python.m) to make sure that python engine is connected to your MATLAB engine as shown in the picture below. 


![Testing python in MATLAB](https://raw.githubusercontent.com/BeamNG/BeamNG-MATLAB-integration/main/media/test_python.png)

<a name="getstart"></a>

## Getting Started

We have developed 5 fundamental scripts that showcase the versatility of using the MATLAB bridge to launch the Beamng.Tech simulator. These scripts demonstrate various capabilities such as utilizing sensors, creating scenarios, capturing screenshots from multiple viewpoints, analyzing movements, and visualizing sensors including lidar, depth camera, and AI driving for the vehicle.

The first script enables the use of sensors, such as lidar and depth camera, to gather data on the vehicle's surroundings. The second script allows for the creation of custom scenarios in the simulator, giving users complete control over the driving environment. The third script facilitates the capturing of screenshots from various angles, allowing for detailed analysis of the vehicle's performance.

The fourth script focuses on movement analysis, providing users with a detailed breakdown of the vehicle's movements and the forces acting on it. Finally, the fifth script provides a visualization of the vehicle's sensor data, including the ability to view the AI driving system in action. Overall, these 5 scripts demonstrate the powerful capabilities of the MATLAB bridge in conjunction with the Beamng.Tech simulator, offering users a comprehensive toolkit for advanced vehicle simulation and analysis.



### 1. Vehicle State Plotting
Use the state sensor to plot some graphs of the vehicle position, wheel speed and direction, throttle, and brake. 

<img src="https://raw.githubusercontent.com/BeamNG/BeamNG-MATLAB-integration/main/media/annotation_bounding_boxes.png" alt="Vehicle state plotting" >

    
### 2. Running Lidar sensor, and AI control. 
Create a simple scenario and use the simulator's AI with BeamNGpy.

<img src="https://raw.githubusercontent.com/BeamNG/BeamNG-MATLAB-integration/main/media/lidar_tour.png" alt="Lidar sensor and AI control mode" >

### 3. Multi-shot Camera    
Change the position and rotation of a camera.

<img src="https://raw.githubusercontent.com/BeamNG/BeamNG-MATLAB-integration/main/media/multi_shots_1.png" alt="Multi-shot Camera" >

### 4. Object Placement    
Define a custom scenario for a given map and generate procedural content, i.e. simple meshes.

<img src="https://raw.githubusercontent.com/BeamNG/BeamNG-MATLAB-integration/main/media/object_placment_0.png" alt="Object Placement" >

### 5. Annotation and Bounding Boxes
Get semantic annotations, instance annotations, and draw bounding boxes (note that this feature is not ready for use yet).

<img src="https://raw.githubusercontent.com/BeamNG/BeamNG-MATLAB-integration/main/media/annotation_bounding_boxes.png" alt="Annotation and Bounding Boxes" >



<a name="compatibility"></a>

## Compatibility  

Running the BeamNG ROS integration requires three individual software components, here is a list of compatible versions.

| BeamNG.tech | BeamNGpy | BeamNG-MATLAB-integration | MATLAB | Python  |
|-------------|----------|---------------------------|--------|---------|
| [0.28](https://beamng.tech/blog/beamng-tech-028/)        | [1.26](https://github.com/BeamNG/BeamNGpy/releases/tag/v1.26)     | [0.1.1](https://github.com/BeamNG/BeamNG-MATLAB-integration/releases/tag/v0.1.1)                     | [R2023a](https://www.mathworks.com/products/new_products/latest_features.html) | [3.9](https://www.python.org/downloads/release/python-390/)  | 
| [0.27](https://beamng.tech/blog/beamng-tech-027/)        | [1.25.1](https://github.com/BeamNG/BeamNGpy/releases/tag/v1.25.1)     | [0.1.0](https://github.com/BeamNG/BeamNG-MATLAB-integration/releases/tag/v0.1.0)                     | [R2022b](https://www.mathworks.com/products/new_products/latest_features.html) | [3.9](https://www.python.org/downloads/release/python-390/)     | 



<a name="license"></a>

## License

This project is licensed under the MIT License - see the [LICENSE](https://github.com/BeamNG/BeamNG-MATLAB-integration/blob/main/LICENSE.txt) file for details.



CHANGELOG.rst
=========
Changelog
=========

Version 0.1.1
===========
- update the code with latest BeamNGpy 1.26 and BeamNG.Tech 0.28


Version 0.1.0
===========
- We have developed Five fundamental scripts that showcase the versatility of using the MATLAB bridge to launch the Beamng.Tech simulator. 
- the Five scripts are: 
    - annotation_bounding_boxes.m
    - lidar_tour.m
    - lidar_tour.m
    - multishot_camera.m
    - object_placement.m
    - vehicle_state_plotting.m

test_python.m
pe = pyenv;
if isempty(pe.Version)
    fprintf ('Python is not installed in your MATLAB engine')
else
    fprintf ('   Your Python version is : ')
    fprintf (pe.Version)
    fprintf('\n')

end


lidar_tour.m
close all
clear
clc

%link bng object to BeamNG folder 
bng =  py.beamngpy.BeamNGpy('localhost', int32(64256));
%bng =  py.beamngpy.BeamNGpy('localhost', int8(64256), home='C:\\game');
%open the BeamNG UI
bng.open();
% Add scenario information; name and level
scenario = py.beamngpy.Scenario('west_coast_usa', 'example');
% Add vehicle's information; name plate no. and vehicle model 
vehicle = py.beamngpy.Vehicle('ego_vehicle', model='etk800', licence='MATLAB');
% specify the veicle's starting position 
scenario.add_vehicle(vehicle, pos=[-717, 101, 118], ...
    rot_quat=[0, 0, 0.3826834, 0.9238795]);
% save all data and load the scenario 
scenario.make(bng);
% waiting for the loaing process
pause(2);
% load the vehicle, level, and scensors in scenario 
bng.load_scenario(scenario);
bng.set_steps_per_second(60);
bng.set_deterministic();
%another wait
pause(2);
%start the scenario 
bng.start_scenario();
%another wait
pause(2);
% adding initial throttle and strearing angle 
vehicle.control(throttle=1);
vehicle.control(steering=0.5);
% setup the AI driving mode 
vehicle.ai_set_mode('span');
lidar1 =  py.beamngpy.sensors.Lidar('lidar1', bng, vehicle, ...
    is_using_shared_memory=true, is_visualised=true);
%another wait

pause(3);
% LiDAR point cloud data (automatic polling)
% parse lidar data in data object 
data = lidar1.poll();

%another wait
pause(1);
                                
% Collect the data now that it has been computed.
bng.step(3, wait=false);


multishot_camera.m

close all
clear
clc

% Start the simulator.
beamng =  py.beamngpy.BeamNGpy('localhost', int32(64256));
beamng.open();

% Create a vehicle.
vehicle = py.beamngpy.Vehicle('ego_vehicle', ...
    pyargs('model', 'etki', 'licence', 'Matlab', 'color', 'Green'));

% Create a scenario.
scenario = py.beamngpy.Scenario('smallgrid', 'multishot', ...
    pyargs('description', 'Demo of the camera sensor used like a multishot camera'));
scenario.add_vehicle(vehicle);
scenario.make(beamng);
beamng.load_scenario(scenario);
beamng.start_scenario();

% Create a default camera sensor, attached to the vehicle.
camera = py.beamngpy.sensors.Camera('camera1', beamng, vehicle);    
% Compute the center of the vehicle bounding box, to give us a point at which to focus the camera.
bbox = vehicle.get_bbox();
rbl = double(py.numpy.array(bbox{'rear_bottom_left'}));
fbr = double(py.numpy.array(bbox{'front_bottom_right'}));
rtl = double(py.numpy.array(bbox{'rear_top_left'}));
center = rbl + ((fbr - rbl) / 2) + ((rtl - rbl) / 2);

 % An offset to move center back a bit to get more of the vehicle.
center(1) = center(1) - 0.5;                                     
 % An offset to move the camera a few metres in front of the center.
pos_offset = center + [5, 0, 0];                                 

% Compute all the angles we will need, in radians.
variations = [ -15 * pi/360, 0.0, 15 * pi/360];

ctr = 1;
cam_images = cell(1, 9);
for h = variations
    for v = variations
        rot_x = cos(h) * pos_offset(1) - sin(h) * pos_offset(2);
        rot_y = sin(h) * pos_offset(1) + cos(h) * pos_offset(2);
        rot_z = sin(v) * rot_y + cos(v) * pos_offset(3);
        new_pos = [rot_x, rot_y, rot_z];
        % Set the camera position for this image.
        camera.set_position(new_pos);                               
        dir = center - new_pos;
        % Set the camera direction for this image.
        camera.set_direction([dir(1), dir(2), dir(3)]);             
         % A slight delay to allow the automated camera to get fresh readings.
        pause(0.3);                                                
        % Get the latest image data from the camera.      
        data = camera.poll();                                       
        % The colour image. We can use {'annotation'} or {'depth'} instead.     
        cam_images{ctr} = double(py.numpy.array(data{'colour'}));   
        ctr = ctr + 1;
    end
end

% Plot all the images.
ctr = 1;
for x = 1:3
    for y = 1:3
        subplot(3, 3, ctr);
        colour_RGBA = double(py.numpy.array(cam_images{ctr}));
        colour_RGB = colour_RGBA(:, :, 1:3) / 256;
        image(colour_RGB);
        ctr = ctr + 1;
    end
end

% Tidy up and close the simulation.
camera.remove();
beamng.close();

object_placement.m
%% 

close all
clear
clc

%% SCENARIO CREATION

% Start the simulator
beamng =  py.beamngpy.BeamNGpy('localhost', int32(64256));
beamng.open();

% Create scenario
scenario = py.beamngpy.Scenario('west_coast_usa', 'object_placement');

% Create a vehicle
vehicle = py.beamngpy.Vehicle('ego_vehicle', pyargs('model', 'etk800', ...
                                                    'licence', 'Matlab', ...
                                                    'color', 'Green'));

scenario.add_vehicle(vehicle, pyargs('pos', py.tuple({-191.62, -160.32, 119.59}), ...
                                     'rot_quat', py.tuple({0, 0, 0.88, -0.4486})));

% Create ramp
ramp = py.beamngpy.StaticObject(pyargs('name', 'pyramp',...
                                       'pos', py.tuple({277.5, 183.5, 118.75}), ...
                                       'rot_quat', py.beamngpy.angle_to_quat(py.tuple({0, 0, 55})), ...
                                       'scale', py.tuple({1, 1, 1}), ...
                                       'shape', '/art/shapes/objects/ramp_massive.dae'));

scenario.add_object(ramp)

% Create ring
ring = py.beamngpy.ProceduralRing(pyargs('name', 'pyring', ...
                                         'pos', py.tuple({445, 301, 230}), ...
                                         'rot_quat', py.beamngpy.angle_to_quat(py.tuple({0, 0, 100})), ...
                                         'radius', 5, ...
                                         'thickness', 2.5));

scenario.add_procedural_mesh(ring)

scenario.make(beamng)
beamng.set_deterministic()
beamng.load_scenario(scenario)

tech_version = true; % Set this to False if you are running BeamNG.drive!

if tech_version
    % Setup camera
    cam_width = 2048;
    cam_height = 2048;
    cam_pos = {391.5, 251, 197.8};
    cam_dir = {445 - cam_pos{1}, 301 - cam_pos{2}, 208 - cam_pos{3}};
    cam_pos = py.tuple(cam_pos);
    cam_dir = py.tuple(cam_dir);
    cam = py.beamngpy.sensors.Camera('camera', beamng, ...
                                     pyargs('pos', cam_pos, ...
                                            'dir', cam_dir, ...
                                            'field_of_view_y', 60, ...
                                            'resolution', py.tuple({int32(cam_width), int32(cam_height)}), ...
                                            'near_far_planes', py.tuple({1, 4000}), ...
                                            'is_static', true, ...
                                            'is_visualised', false));
end

%% RUNNING
beamng.start_scenario()

meshes = scenario.find_procedural_meshes();
meshes = cell(meshes);

% Find ring position
for ii = 1:length(meshes)
    mesh = meshes{ii};
    if strcmp(char(mesh.name), 'pyring')
        ring_pos = double(mesh.pos);
    end
end

% Move vehicle
for ii = 1:5
    vehicle.control(pyargs('throttle', ii/4))
    beamng.step(int32(150))
end

% Compute distance between vehicle and ring
scenario.update()
distance = norm(double(vehicle.state{'pos'}) - ring_pos);

while distance > 5
    scenario.update()
    distance = norm(double(vehicle.state{'pos'}) - ring_pos);
end

% Get camera images
if tech_version
    frames = cam.poll();
end

beamng.close()

% Plot images
colour_RGBA = double(py.numpy.array(frames{'colour'}));
colour_RGB = colour_RGBA(:, :, 1:3) / 256;
image(colour_RGB);
axis equal
xlim([0 cam_width])
ylim([0 cam_height])



vehicle_state_plotting.m
%% 

close all
clear
clc

%% SCENARIO CREATION

% Start the simulator
beamng =  py.beamngpy.BeamNGpy('localhost', int32(64256));
beamng.open();

% Create a vehicle
vehicle = py.beamngpy.Vehicle('ego_vehicle', pyargs('model', 'etk800', ...
                                                    'licence', 'Matlab', ...
                                                    'color', 'Green'));

% Create electrics sensor
electrics = py.beamngpy.sensors.Electrics();
vehicle.attach_sensor('electrics', electrics)

% Create scenario
scenario = py.beamngpy.Scenario('west_coast_usa', 'vehicle_state');

scenario.add_vehicle(vehicle, pyargs('pos', py.tuple({-717.121, 101, 118.675}), ...
                                     'rot_quat', py.beamngpy.angle_to_quat(py.tuple({0, 0, 45}))));

scenario.make(beamng)

%% RUNNING

beamng.load_scenario(scenario)
beamng.start_scenario()

vehicle.ai_set_mode('span')

% Initialize output array
n_sample = 240;

positions = zeros(n_sample, 3);
directions = zeros(n_sample, 3);
wheel_speeds = zeros(n_sample, 1);
throttles = zeros(n_sample, 1);
brakes = zeros(n_sample, 1);

% Display sensor output
vehicle.poll_sensors()
sensors = vehicle.sensors;

position_init = vehicle.state{'pos'};
direction_init = vehicle.state{'dir'};
electrics_sensor = sensors.data{'electrics'};
wheel_speed_init = electrics_sensor{'wheelspeed'};
throttle_init = electrics_sensor{'throttle'};
brake_init = electrics_sensor{'brake'};

disp(strcat("The vehicle position is: ", num2str(double(position_init))))
disp(strcat("The vehicle direction is: ", num2str(double(direction_init))))
disp(strcat("The wheel speed is: ", num2str(wheel_speed_init)))
disp(strcat("The throttle intensity is: ", num2str(throttle_init)))
disp(strcat("The brake intensity is: ", num2str(brake_init)))


% Acquire data
for ii = 1:n_sample
    pause(0.1)
    vehicle.poll_sensors()
    sensors = vehicle.sensors;
    electrics_sensor = sensors.data{'electrics'};
    positions(ii, :) = vehicle.state{'pos'};
    directions(ii, :) = vehicle.state{'dir'};
    wheel_speeds(ii) = electrics_sensor{'wheelspeed'};
    throttles(ii) = electrics_sensor{'throttle'};
    brakes(ii) = electrics_sensor{'brake'};
end

beamng.close()

%% PLOTTING
subplot(2, 2, [1 2])
plot(throttles, 'b')
hold on
plot(brakes, 'r')
grid on
legend('Throttle', 'Brake')
title('Brake and Throttle vs time')

subplot 223
plot(positions(:, 1), positions(:, 2))
grid on
title('Position')
xlabel('X')
ylabel('Y')

subplot 224
angles = atan2(directions(:, 2), directions(:, 1));
polarplot(angles, wheel_speeds)
grid on
title('Wheel Speed and Direction')


annotation_bounding_boxes.m

close all
clear
clc

% Executing this file will demonstrate the annotation functionality of the camera sensor.

% Link bng object to BeamNG folder.
%bng =  py.beamngpy.BeamNGpy('localhost', int8(64256), home='C:\\Users\\User\\AppData\\BeamNG.drive');
bng =  py.beamngpy.BeamNGpy('localhost', int32(64256));
bng.open()                                                       % Start up the simulator.
scenario = py.beamngpy.Scenario('italy', 'annotation_bounding_boxes');      % Create the scenario.
%
% Create some vehicles which will appear in the camera images.
ego = py.beamngpy.Vehicle('ego', pyargs('model', 'etk800', 'color', 'White'));
scenario.add_vehicle(ego, pyargs('pos', py.list({237.90, -894.42, 246.10}), 'rot_quat', py.list({0.0173, -0.0019, -0.6354, 0.7720})));
car1 = py.beamngpy.Vehicle('car1', pyargs('model', 'etk800', 'color', 'Green'));
scenario.add_vehicle(car1, pyargs('pos', py.list({246.94, -901.64, 247.58}), 'rot_quat', py.list({-0.0099, 0.0206, 0.9348, -0.3543})));
car2 = py.beamngpy.Vehicle('car2', pyargs('model', 'etk800', 'color', 'Red'));
scenario.add_vehicle(car2, pyargs('pos', py.list({276.27, -881.42, 247.84}), 'rot_quat', py.list({-0.0106, 0.0405, 0.4845, 0.8738})));
car3 = py.beamngpy.Vehicle('car3', pyargs('model', 'etki', 'color', 'Blue'));
scenario.add_vehicle(car3, pyargs('pos', py.list({261.52, -894.68, 248.04}), 'rot_quat', py.list({0.0026, 0.01758, -0.8344, 0.5508})));
car4 = py.beamngpy.Vehicle('car4', pyargs('model', 'miramar', 'color', 'Black'));
scenario.add_vehicle(car4, pyargs('pos', py.list({267.06, -892.03, 248.32}), 'rot_quat', py.list({0.0065, 0.0194, -0.8501, 0.5262})));

scenario.make(bng);
bng.set_deterministic();
bng.set_steps_per_second(int32(60));                           % Set simulator to 60hz temporal resolution
bng.load_scenario(scenario);
bng.hide_hud();
bng.start_scenario();
bng.switch_vehicle(ego);

% Get the annotation class data.
annotations = bng.get_annotations();                    % Gets a dictionary of RGB colours, indexed by material names.
disp(annotations);
class_data = bng.get_annotation_classes(annotations);   % Gets a dictionary of material names, indexed by RGB colours (encoded as 32-bit).
disp(class_data);
pause(int32(10));                                              % Some sleeping time to make sure the level is fully loaded.

% Create some camera sensors in the simulation.
disp('Camera test start.');

% Create a camera sensor.
cam_width = 1024;
cam_height = 1024;

% args = pyargs( ...
%     'requested_update_time', -1.0, ...
%     'is_render_instance', true, ...
%     'pos', py.list({-0.3, 1.0, 2.0}), ...
%     'resolution', py.list({int32(cam_width), int32(cam_height)}));


args = pyargs( ...
            'requested_update_time', -1.0, ...
              'is_using_shared_memory', false, ...
                  'pos', py.list({-0.3, 1.0, 2.0}), ...
                  'dir', py.list({0, -1.0, 0.0}), ...
                  'field_of_view_y' , 70 , ...
                  'near_far_planes' ,  py.list({0.1, 100}), ... 
                 'resolution', py.list({int32(cam_width), int32(cam_height)}), ...
                  'is_render_annotations', true, ...
                  'is_render_instance', true, ...
                  'is_render_depth', true);

cam1 = py.beamngpy.sensors.Camera('camera1', bng, ego, args);

% Wait a couple of seconds (to give camera sensor time to update) then poll it.
pause(2);
images = cam1.get_full_poll_request();

% Plot the camera images.
subplot 231
colour_RGBA = double(py.numpy.array(images{'colour'}));
colour_RGB = colour_RGBA(:, :, 1:3) / 256;
image(colour_RGB);
title('Colour');
axis equal
xlim([0 cam_width])
ylim([0 cam_height])

subplot 232
annotation_RGBA = double(py.numpy.array(images{'annotation'}));
annotation_RGB = annotation_RGBA(:, :, 1:3) / 256;
image(annotation_RGB);
title('Class Annotations');
axis equal
xlim([0 cam_width])
ylim([0 cam_height])

subplot 233
instance_RGBA = double(py.numpy.array(images{'instance'}));
instance_RGB = instance_RGBA(:, :, 1:3) / 256;
image(instance_RGB);
title('Instance Annotations');
axis equal
xlim([0 cam_width])
ylim([0 cam_height])

subplot 234
depth_RGB = double(py.numpy.array(images{'depth'}));
image(depth_RGB);
colormap(gray)
title('Depth');
axis equal
xlim([0 cam_width])
ylim([0 cam_height])

% Compute and display the bounding boxes for each vehicle instance.
bounding_boxes = cam1.extract_bounding_boxes(images{'annotation'}, images{'instance'}, class_data);
for i = bounding_boxes.length - 1 : -1 : 1
    n = bounding_boxes(i);
    disp(n{1}{'class'})
    if n{1}{'class'} ~= "CAR"
        bounding_boxes.remove(n{1});
    end
end
disp(bounding_boxes);

% Draw the bounding boxes on top of the colour image.
image_with_boxes = cam1.draw_bounding_boxes(bounding_boxes, images{'colour'});
subplot 235
image_with_boxes_RGBA = double(py.numpy.array(image_with_boxes));
image_with_boxes_RGB = image_with_boxes_RGBA(:, :, 1:3) / 256;
image(image_with_boxes_RGB);
title('Bounding Boxes');
axis equal
xlim([0 cam_width])
ylim([0 cam_height])

% Export the bounding boxes to PASCAL-VOC XML format.
xml = cam1.export_bounding_boxes_xml(bounding_boxes, pyargs('filename', 'example.png', ...
                                                            'size', py.list({cam_width, cam_height, 3})));
disp(xml);

% Close the simulation.
bng.close();
